edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/modules/MqttBrokerModule.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Service.Modules
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Autofac;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+
+    class MqttBrokerModule : Module
+    {
+        static readonly int defaultPort = 1883;
+        static readonly string defaultUrl = "127.0.0.1";
+
+        readonly IConfiguration config;
+
+        public MqttBrokerModule(IConfiguration config)
+        {
+            this.config = Preconditions.CheckNotNull(config, nameof(config));
+        }
+
+        protected override void Load(ContainerBuilder builder)
+        {
+            var componentTypes = MqttBridgeComponentDiscovery.GetCandidateTypes().ToArray();
+
+            // The classes will be registered by two types:
+            // 1) by its own type - that is because the connector is going to
+            //    request an instance by the type name as it finds ISubscriber, etc
+            //    derived types.
+            // 2) by every interface name that is not the following:
+            //      IMessageConsumer/IProducer/ISubscriber/ISubscriptionWatcher.
+            //    That is because some of the producers/consumers are also special services
+            //    and components will need those injected by custom interfaces.
+            builder.RegisterTypes(componentTypes)
+                   .AsSelf()
+                   .As(t => GetNonStandardBridgeInterfaces(t))
+                   .SingleInstance();
+
+            builder.RegisterType<DeviceProxy>()
```

Comment:
Why do we need this?

Reply:
The registrations happen for the following uses:
- the component discovery will collect all types implementing IMessageConsumer, etc, then resolves them by type (e.g. it has its in list that TwinHandler implements IMessageConsumer, so it will ask autofac for a TwinHandler instance. That is why these types are registered "AsSelf" (e.g. as TwinHandler)
- there are types that use these classes (like TwinHandler) as special services. They know them from an interface, like ITwinHandler (e.g the DeviceProxy class gets ITwinHander injected. That is why every class is registered all of their implemented interfaces, which is not the usual "IMessageConsumer", etc. A class cannot be registered by "IMessageConsumer" because there are several classes implementing them and it would be ambiguous to resolve by them. What happens here is the the "GetNosStandardBridgeInterfaces()" filters out "IMessageConsumer", but leaves there eg "ITwinHandler"

As a result, every "normal" dependency injection use cases (like a constructor parameter receiving an ITwinHandler) can be resolved, but the discovery class also can resolve them by specific type (e.g TwinHandler, because discovery will see that the TwinHandler class is  one implementing IMessageConsumer, so it will request for TwinHandler to put it into its "consumers" list)